[dtensor] support creating DTensor in submesh

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95655
* __->__ #95458
* #95457
* #95456

This PR supports creating DTensor in a submesh, if the rank is not
participating in the mesh, we assign the local tensor to be empty
tensor, and do nothing in the operator dispatch

Differential Revision: [D43643577](https://our.internmc.facebook.com/intern/diff/D43643577)

wanchaol committed:
[dtensor] support creating DTensor in submesh

This PR supports creating DTensor in a submesh, if the rank is not
participating in the mesh, we assign the local tensor to be empty
tensor, and do nothing in the operator dispatch

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] support creating DTensor in submesh"


This PR supports creating DTensor in a submesh, if the rank is not
participating in the mesh, we assign the local tensor to be empty
tensor, and do nothing in the operator dispatch

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] support creating DTensor in submesh"


This PR supports creating DTensor in a submesh, if the rank is not
participating in the mesh, we assign the local tensor to be empty
tensor, and do nothing in the operator dispatch

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -129,21 +148,32 @@ def operator_dispatch(
-    local_tensor_args = cast(Tuple[object, ...], local_tensor_args)
-    local_tensor_kwargs = cast(Dict[str, object], local_tensor_kwargs)
-    local_results = op_call(*local_tensor_args, **local_tensor_kwargs)
+    if mesh is not None and mesh.get_coordinate() is None:
+        # if we are on a non-participating device, we simply return
+        # an empty tensor for now.
+        # TODO: what if the op returns a non-tensor value, what if
+        # the op returns a list of tensors, we need to figure out
+        # a consistent way to handle that, and also need to figure
+        # out if we should communicate the result to non-participating
+        # ranks (i.e. a.sum() -> scalar, maybe we should set to 0)
+        local_results = torch.tensor([])
```
1. Communicating the result to non-participating ranks will block those ranks so we may not want to do it unless necessary. If we really want some result on those ranks, we can use a default value passed by users but that requires a change of all DTensor ops argument list by adding a default value.

2. The op return type problem should also be fixed by using a default value passed in.

wanchaol replied:
I think we should set the default values by ourselves, I will submit a follow up PR later to address this.

XilunWu reviewed (approved):
lgtm!

wanchaol committed:
Update on "[dtensor] support creating DTensor in submesh"


This PR supports creating DTensor in a submesh, if the rank is not
participating in the mesh, we assign the local tensor to be empty
tensor, and do nothing in the operator dispatch

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43643577).

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -f "failure not related"
